### DISCUSSION TITLE
Popper element without Portal

### DISCUSSION THREAD

**Discussion by jvdsande** [mantinedev/mantine#435] [Q&A] [answered]
Hello again

Would it be possible to make the `Portal` element inside `Popper` optional? 

I get that most of the time for a Popover or a Tooltip this would be preferred, but now that it is used for the Select too, it creates some unwanted behaviors. Being able to have the Popper element of the Select located at the same level in the Dom would be better.

Here are a few things that come to mind where a Popper element located next to the Select would be good:

- Inheriting CSS variables
- z-index positioning: with a Portal our dropdown is now completely separated from the Z layer of the Select itself. This means that it goes above things it should not, for instance here is a screenshot from the Mantine docs, where the dropdown goes above the top-bar
![image](https://user-images.githubusercontent.com/12484011/142674672-e21ec972-8b99-445a-956c-1d37cd4f6ad1.png)

**Comment by rtivital** [comment, marked as answer]
Hi @jvdsande,

> Inheriting CSS variables

Unfortunately, this is the whole idea of Popper component – put dropdown outside of current styles context, this includes CSS variables as well. If this is not done, then there will be z-index issues (we had them before migration to portals).

> z-index positioning

This is another issue of default values, we should configure dropdowns to have smaller dropdown z-index and provide an option to configure that. 

Anyway, if you have other points, I can consider adding this feature.

**Reply by jvdsande** [reply]
Well, Popper.js _does_ allow for a target that is collocated with the reference, so I think it could be good to have this option in Mantine's Popper too.

I have a use case where I need to style the Select component and react to its surroundings. Simplified you could say that I have the Dark theme toggled only on part of the application (that's not Mantine's dark theme here, it's an external dark theme that I have to hook into)

Now this dark theme is propagating values through CSS variables. Unfortunately Popper's Portal lies outside of the root on which the variables are set. I don't see any way I can make it work like this, and it forces me to freeze versions to 3.1.9

I totally get that using a Portal like so is a great default for the built-in components, but it plays against the otherwise awesome customizability of Mantine that users cannot change that default.

In the Select component, we have the option to introduce our own `dropdownComponent` but unfortunately this is already one level too deep, as it is the content of the Popper that gets replaced, not the Popper itself.

**Reply by jvdsande** [reply]
In my understanding the primary goal of Popper component is to offer placement behavior, with out-of-bounds detection for shifting the placement. This can be achieved on a target other than a Portal.

For the zindex issues you refer too, I guess those where more linked to Popover and Tooltip components, and not Select though? Since Select neither used Popper or a Portal until now.

From a design standpoint, I really feel like the drop-down of a select is inherently part of the element, not like in a menu where it's an action triggering the display of a modal. In a Menu you could imagine a backdrop for instance, not so much in a Select.

I feel like the inclusion of Select in the consumers of Popper means that Popper customization options need to be adapted in consequence.

**Reply by rtivital** [reply]
Okay, I got your use case, I'll add this customization to 3.3

**Reply by jvdsande** [reply]
Thanks, that's awesome to hear!

Let me know if you want any help on it, I can totally do the PR to add this if you want.

Mantine's a really awesome piece of software, thank you for it, and thanks for the very fast responses 🙂

**Reply by rtivital** [reply]
Thanks! I think you can help with it, it seems pretty easy. My initial plan for this was:
- Add `disabled` prop to `Portal` component, which will determine if its content will be rendered with or without `React.createPortal`
- Then just add `withinPortal` prop to Select, Autocomplete, MultiSelect, Menu, Popover and Tooltip components
- `withinPortal` should default to `true` in all components
- Update documentation to reflect changes

If you want to work on this, please submit a PR

**Reply by jvdsande** [reply]
Plan sounds good, I'll work on this tomorrow I think, and open the PR
